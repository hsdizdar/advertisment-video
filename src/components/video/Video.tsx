import { FC, useRef, useEffect, useState } from "react";

interface VideoProps {
  source: string;
}

const Video: FC<VideoProps> = ({ source }) => {
  const videoRef = useRef<HTMLVideoElement>(null);
  const [duration, setDuration] = useState<number>(0);

  useEffect(() => {
    console.log("useEffect");

    const options = {
      rootMargin: "0px",
      threshold: 0.5,
    };

    const handlePlay = (entries: Array<IntersectionObserverEntry>) => {
      console.log("handlePlay: ", entries);

      entries.forEach((entry) => {
        console.log("handlePlay forEach: ", entry.isIntersecting);

        if (entry.isIntersecting) {
          console.log("autoPlay: ", videoRef.current?.autoplay);
          videoRef.current?.play();
        } else {
          videoRef.current?.pause();
        }
      });
    };

    const observer = new IntersectionObserver(handlePlay, options);

    console.log("observer: ", observer);
    console.log("videoRef.current: ", videoRef.current);

    observer.observe(videoRef.current as Element);
  }, []);

  const handleDuration = () => {
    const currentTime = videoRef?.current?.currentTime!;

    const calc = (currentTime / duration).toFixed(2);

    if (calc === "0.25") {
      console.log("%25");
    } else if (calc === "0.50") {
      console.log("%50");
    } else if (calc === "0.75") {
      console.log("%75");
    }
  };

  const handleLoad = () => {
    const duration = videoRef.current?.duration;
    setDuration(duration || 0);
  };

  const handleEnd = () => {
    console.log("Video is ended.");
  };

  return (
    <video
      style={{ width: "100%" }}
      ref={videoRef}
      src={source}
      onTimeUpdate={handleDuration}
      onEnded={handleEnd}
      onLoadedData={handleLoad}
      data-testid="video"
      controls
      muted
    />
  );
};

export default Video;
