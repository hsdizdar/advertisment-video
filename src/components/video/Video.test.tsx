import { act } from "react-dom/test-utils";
import userEvent from "@testing-library/user-event";

import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";

import Video from "./Video";

describe("useIntersectionObserver", () => {
  beforeEach(() => {
    const mockIntersectionObserver = jest.fn();

    mockIntersectionObserver.mockReturnValue({
      observe: () => jest.fn(),
      thresholds: 0.5,
    });

    window.IntersectionObserver = mockIntersectionObserver;
  });

  const source = "https://cdn.yoc.com/ad/demo/airbnb.mp4";

  it("should render video", async () => {
    const { container } = render(<Video source={source} />);
    const video = screen.getByTestId("video");
    expect(video).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should play video when clicked", async () => {
    const { container } = render(<Video source={source} />);
    const video = screen.getByTestId("video");
    expect(video).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
    act(() => {
      userEvent.click(video);
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should video end", async () => {
    const { container } = render(<Video source={source} />);
    const video = screen.getByTestId("video");
    expect(video).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
    act(() => {
      userEvent.click(video);
    });
    expect(container.firstChild).toMatchSnapshot();
    act(() => {
      video.dispatchEvent(new Event("ended"));
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should play video when appears 50%", async () => {
    const { container } = render(<Video source={source} />);
    const video = screen.getByTestId("video");
    expect(video).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toMatchSnapshot();
    act(() => {
      video.dispatchEvent(new Event("isIntersection"));
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
