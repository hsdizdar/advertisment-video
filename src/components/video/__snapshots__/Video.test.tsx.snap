// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useIntersectionObserver should play video when appears 50% 1`] = `
<video
  controls=""
  data-testid="video"
  src="https://cdn.yoc.com/ad/demo/airbnb.mp4"
  style="width: 100%;"
/>
`;

exports[`useIntersectionObserver should play video when appears 50% 2`] = `
<video
  controls=""
  data-testid="video"
  src="https://cdn.yoc.com/ad/demo/airbnb.mp4"
  style="width: 100%;"
/>
`;

exports[`useIntersectionObserver should play video when appears 50% 3`] = `
<video
  controls=""
  data-testid="video"
  src="https://cdn.yoc.com/ad/demo/airbnb.mp4"
  style="width: 100%;"
/>
`;

exports[`useIntersectionObserver should play video when clicked 1`] = `
<video
  controls=""
  data-testid="video"
  src="https://cdn.yoc.com/ad/demo/airbnb.mp4"
  style="width: 100%;"
/>
`;

exports[`useIntersectionObserver should play video when clicked 2`] = `
<video
  controls=""
  data-testid="video"
  src="https://cdn.yoc.com/ad/demo/airbnb.mp4"
  style="width: 100%;"
/>
`;

exports[`useIntersectionObserver should render video 1`] = `
<video
  controls=""
  data-testid="video"
  src="https://cdn.yoc.com/ad/demo/airbnb.mp4"
  style="width: 100%;"
/>
`;

exports[`useIntersectionObserver should video end 1`] = `
<video
  controls=""
  data-testid="video"
  src="https://cdn.yoc.com/ad/demo/airbnb.mp4"
  style="width: 100%;"
/>
`;

exports[`useIntersectionObserver should video end 2`] = `
<video
  controls=""
  data-testid="video"
  src="https://cdn.yoc.com/ad/demo/airbnb.mp4"
  style="width: 100%;"
/>
`;

exports[`useIntersectionObserver should video end 3`] = `
<video
  controls=""
  data-testid="video"
  src="https://cdn.yoc.com/ad/demo/airbnb.mp4"
  style="width: 100%;"
/>
`;
